src/internal/ajax/types.ts

```diff
@@ -166,12 +166,33 @@ export interface AjaxConfig {
   createXHR?: () => XMLHttpRequest;
 
   /**
-   * An observer for watching the progress of an HTTP request. Will
-   * emit progress events, and completes on the final load event, will error for
+   * An observer for watching the upload progress of an HTTP request. Will
+   * emit progress events, and completes on the final upload load event, will error for
    * any XHR error or timeout.
    *
    * This will **not** error for errored status codes. Rather, it will always _complete_ when
    * the HTTP response comes back.
+   *
+   * @deprecated If you're looking for progress events, please try {@link includeDownloadProgress} and
+   * {@link includeUploadProgress}. This will be removed in version 8.
    */
   progressSubscriber?: PartialObserver<ProgressEvent>;
+
+  /**
+   * If `true`, will emit all download progress and load complete events as {@link AjaxProgressEvents}
+   * from the observable. The final download event will also be emitted as a {@link AjaxDownloadCompleteEvent}
+   *
+   * If both this and {@link includeUploadProgress} are `false`, then only the {@link AjaxResponse} will
+   * be emitted from the resulting observable.
+   */
+  includeDownloadProgress?: boolean;
+
+  /**
+   * If `true`, will emit all upload progress and load complete events as {@link AjaxUploadProgressEvents}
+   * from the observable. The final download event will also be emitted as a {@link AjaxDownloadCompleteEvent}
+   *
+   * If both this and {@link includeDownloadProgress} are `false`, then only the {@link AjaxResponse} will
+   * be emitted from the resulting observable.
+   */
+  includeUploadProgress?: boolean;
```

Comment:
When this option is `true`, shouldn't `ajax` return `Observable< AjaxUploadProgressEvents | AjaxDownloadCompleteEvent | AjaxResponse<T>>` instead of `Observable<AjaxResponse<T>>`? Likewise for `includeDownloadProgress` and `AjaxProgressEvents`. /cc @benlesh

Reply:
https://github.com/ReactiveX/rxjs/issues/6145